internal/dataplane/parser/translate_routes_helpers.go

```diff
@@ -0,0 +1,183 @@
+	case gatewayv1alpha2.UDPRouteRule:
+		if len(tt.BackendRefs) == 0 {
+			return nil, errors.New("UDPRoute rules must include at least one backendRef")
+		}
+		return tt.BackendRefs, nil
+	case gatewayv1alpha2.TCPRouteRule:
+		if len(tt.BackendRefs) == 0 {
+			return nil, errors.New("TCPRoute rules must include at least one backendRef")
+		}
+		return tt.BackendRefs, nil
+	case gatewayv1alpha2.TLSRouteRule:
+		// TLSRoutes don't require BackendRefs.
+		return tt.BackendRefs, nil
+	}
+
+	// This should never happen because we use type constraints on what types
+	// are accepted.
+	return nil, nil
+}
+
+// generateKongRoutesFromRouteRule converts a Gateway Route (TCP, UDP or TLS) rule
+// to one or more Kong Route objects to route traffic to services.
+func generateKongRoutesFromRouteRule[T tRoute, TRule tRouteRule](
+	route T,
+	ruleNumber int,
+	rule TRule,
+) ([]kongstate.Route, error) {
+	backendRefs, err := getBackendRefs(rule)
+	if err != nil {
+		return []kongstate.Route{}, err
+	}
+
+	return []kongstate.Route{
+		{
+			Ingress: util.FromK8sObject(route),
+			Route:   routeToKongRoute(route, backendRefs, ruleNumber),
+		},
+	}, nil
+}
+
+// routeToKongRoute converts Gateway Route to kong.Route.
+func routeToKongRoute[TRoute tTCPorUDPorTLSRoute](
+	r TRoute,
+	backendRefs []gatewayv1alpha2.BackendRef,
+	ruleNumber int,
+) kong.Route {
+	switch rr := any(r).(type) {
+	case *gatewayv1alpha2.UDPRoute:
+		return udpRouteToKongRoute(rr, backendRefs, ruleNumber)
+	case *gatewayv1alpha2.TCPRoute:
+		return tcpRouteToKongRoute(rr, backendRefs, ruleNumber)
+	case *gatewayv1alpha2.TLSRoute:
+		return tlsRouteToKongRoute(rr, ruleNumber)
+	}
+
+	return kong.Route{}
+}
+
+func udpRouteToKongRoute(
+	r *gatewayv1alpha2.UDPRoute,
+	backendRefs []gatewayv1alpha2.BackendRef,
+	ruleNumber int,
+) kong.Route {
+	return kong.Route{
+		Name: kong.String(
+			generateRouteName(udpRouteType, r.Namespace, r.Name, ruleNumber)),
+		Protocols:    kong.StringSlice("udp"),
+		Destinations: backendRefsToKongCIDRPorts(backendRefs),
+	}
+}
+
+func tcpRouteToKongRoute(
+	r *gatewayv1alpha2.TCPRoute,
+	backendRefs []gatewayv1alpha2.BackendRef,
+	ruleNumber int,
+) kong.Route {
+	return kong.Route{
+		Name: kong.String(
+			generateRouteName(tcpRouteType, r.Namespace, r.Name, ruleNumber)),
+		Protocols:    kong.StringSlice("tcp"),
+		Destinations: backendRefsToKongCIDRPorts(backendRefs),
+	}
+}
+
+func backendRefsToKongCIDRPorts(backendRefs []gatewayv1alpha2.BackendRef) []*kong.CIDRPort {
+	// For now, Gateway Routes provide no means of specifying a destination port
+	// other than the backend target port
+	//
+	// They will once https://gateway-api.sigs.k8s.io/geps/gep-957/ is stable.
+	// In the interim, this always uses the backend target.
+	//
+	// NOTE: The above is now implemented via:
+	// https://github.com/kubernetes-sigs/gateway-api/pull/1002 and here's the
+	// related issue in KIC:
+	// https://github.com/Kong/kubernetes-ingress-controller/issues/2709
```

Comment:
I recommend explicitly adding `TODO` to the comment, as we have always historically done this for issue links and it's how I search for them in our repos:

```suggestion
	//
	// TODO: https://github.com/Kong/kubernetes-ingress-controller/issues/2709
```